Meta/serenity.sh

```diff
@@ -0,0 +1,190 @@
+    run:       TARGET lagom:      Runs the unit tests on the build host
+               All other TARGETs: $ARG0 run [TARGET] [kernel_cmd_line]
+                   Runs the built image in QEMU, and optionally passed the
+    delete:    Removes the build environment for the TARGET
+    recreate:  Deletes and re-creates the build environment for TARGET
+    rebuild:   Deletes and re-creates the build environment, and compiles for TARGET
+    addr2line: $ARG0 addr2line TARGET BINARY_FILE ADDRESS
+                   Resolves the ADDRESS in BINARY_FILE to a file:line
+
+    rebuild-toolchain: Deletes and re-builds the TARGET's toolchain
+  Supported TARGETs: i686 (default), x86_64, lagom
+
+  Examples:
+    $ARG0 run i686 smp=on
+        Runs the image in QEMU passing "smp=on" to the kernel command line
+    $ARG0 run
+        Runs the image for the default TARGET i686 in QEMU
+    $ARG0 run lagom
+        Runs the unit tests on the build host
+    $ARG0 addr2line i686 Kernel/Kernel 0x12345678
+        Resolves the address 0x12345678 in the binary Kernel/Kernel
+EOF
+}
+
+function usage()
+{
+  >&2 print_help
+  exit 1
+}
+
+CMD=$1;
+[ ! -z "$CMD" ] || usage
+shift
+if [ $CMD = "help" ]; then
+  print_help
+  exit 0
+fi
+
+if [ ! -z "$1" ]; then
+  TARGET="$1"; shift
+else
+  TARGET="i686"
+fi
+CMD_ARGS=( "$@" )
+CMAKE_ARGS=()
+
+function get_top_dir()
+{
+  git rev-parse --show-toplevel
+}
+
+function is_valid_target()
+{
+  if [ $TARGET = "lagom" ]; then
+    echo 'target is lagom';
+    CMAKE_ARGS+=("-DBUILD_LAGOM=ON")
+    return 0
+  fi
+  [[ $TARGET =~ ^(i686|x86_64|lagom)$ ]] || return 1
+}
+
+function create_build_dir()
+{
+  mkdir -p "$BUILD_DIR"
+  ( cd "$BUILD_DIR" && cmake -GNinja "${CMAKE_ARGS[@]}" ../..)
+}
+
+function cmd_with_target()
+{
+  is_valid_target || ( >&2 echo "Unknown target: $TARGET"; usage )
+  export SERENITY_ARCH="$TARGET"
+  SERENITY_ROOT=$(get_top_dir)
+  export SERENITY_ROOT
+  BUILD_DIR="$SERENITY_ROOT/Build/$TARGET"
+  if [ $TARGET = "lagom" ]; then
+    TOOLCHAIN=i686
+  else
+    TOOLCHAIN="$TARGET"
+  fi
+  TOOLCHAIN_DIR="$SERENITY_ROOT/Toolchain/Build/$TARGET"
+}
+
+function ensure_target()
+{
+  [ -d "$BUILD_DIR" ] || create_build_dir
+}
+
+function build_target()
+{
+  ( cd "$BUILD_DIR" && ninja "$@" )
+}
+
+function delete_target()
+{
+  [ ! -d "$BUILD_DIR" ] || rm -rf "$BUILD_DIR"
+}
+
+function build_toolchain()
+{
+  ( cd Toolchain && ARCH="$TOOLCHAIN" ./BuildIt.sh )
+}
+
+function ensure_toolchain()
+{
+  [ -d "$TOOLCHAIN_DIR" ] || build_toolchain
+}
+
+function delete_toolchain()
+{
+  [ ! -d "$TOOLCHAIN_DIR" ] || rm -rf "$TOOLCHAIN_DIR"
+}
+
+function not_supported_lagom()
+{
+  >&2 echo "Command '$CMD' not supported for the lagom target"
+  exit 1
+}
+
+if [[ $CMD =~ ^(build|install|image|run|rebuild|recreate|addr2line|setup-and-run)$ ]]; then
+  cmd_with_target
+  [ $CMD != "recreate" -a $CMD != "rebuild" ] || delete_target
+  ensure_target
+  ensure_toolchain
```

Comment:
If the toolchain doesn't exist yet, I think this will go wrong: Telling cmake to use a compiler that doesn't exist yet sounds like a bad idea.

Suggestion:
```
ensure_toolchain
ensure_target
```
Or does this cause other issues?

Reply:
I'll have to check it out. When I wrote my scripts I had to do it this way or the BuildIt.sh of the toolchain would run cmake for me or something weird like this. Maybe it's no longer necessary.

Reply:
Changing it to the more correct order somehow causes this error after building the toolchain:

```
ninja: error: loading 'build.ninja': No such file or directory
```

Not sure why the other order works, maybe it's picking up the wrong compiler indeed.

Reply:
So it appears to work in the wrong order because it doesn't look like cmake is actually looking for the toolchain. When I have the toolchain built and manually run cmake after, I get this:
```
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
```
Why it doesn't work in the correct order, I'm not sure. I've been using the wrong (though working) order for a long, long time, so I think for now I'll just leave it as is.